src/controllers/version/metadata.rs

```diff
@@ -84,3 +104,145 @@ pub async fn show(
+///
+/// This endpoint allows updating the yanked state of a version, including a yank message.
+pub async fn update(
+    state: AppState,
+    Path((crate_name, version)): Path<(String, String)>,
+    req: Parts,
+    Json(update_data): Json<VersionUpdate>,
+) -> AppResult<Json<Value>> {
+    if semver::Version::parse(&version).is_err() {
+        return Err(version_not_found(&crate_name, &version));
+    }
+
+    let conn = state.db_write().await?;
+    spawn_blocking(move || {
+        let conn: &mut AsyncConnectionWrapper<_> = &mut conn.into();
+        let (mut version, krate) = version_and_crate(conn, &crate_name, &version)?;
+
+        apply_yank_update(&state, &req, conn, &mut version, &krate, &update_data)?;
+
+        let published_by = version.published_by(conn);
+        let actions = VersionOwnerAction::by_version(conn, &version)?;
+        let updated_version = EncodableVersion::from(version, &krate.name, published_by, actions);
+        Ok(Json(json!({ "version": updated_version })))
+    })
+    .await
+}
+
+fn apply_yank_update(
+    state: &AppState,
+    req: &Parts,
+    conn: &mut impl Conn,
+    version: &mut Version,
+    krate: &Crate,
+    update_data: &VersionUpdate,
+) -> AppResult<()> {
+    // Try to update the yank state first, to avoid unnecessary checks.
+    update_version_yank_state(version, update_data)?;
+    perform_version_yank_update(state, req, conn, version, krate)?;
+
+    Ok(())
+}
+
+fn update_version_yank_state(version: &mut Version, update_data: &VersionUpdate) -> AppResult<()> {
+    match (update_data.yanked, &update_data.yank_message) {
+        (Some(true), Some(message)) => {
+            version.yanked = true;
+            version.yank_message = Some(message.clone());
+        }
+        (Some(yanked), None) => {
+            version.yanked = yanked;
+            version.yank_message = None;
+        }
+        (Some(false), Some(_)) => {
+            return Err(bad_request("Cannot set yank message when unyanking"));
+        }
+        (None, Some(message)) => {
+            if version.yanked {
+                version.yank_message = Some(message.clone());
+            } else {
+                return Err(bad_request(
+                    "Cannot update yank message for a version that is not yanked",
+                ));
+            }
+        }
+        // If both yanked and yank_message are None, do nothing.
+        // This function only cares about updating the yanked state and yank message.
+        (None, None) => {}
+    }
+    Ok(())
+}
+
+pub fn perform_version_yank_update(
+    state: &AppState,
+    req: &Parts,
+    conn: &mut impl Conn,
+    version: &Version,
+    krate: &Crate,
+) -> AppResult<()> {
+    let auth = AuthCheck::default()
+        .with_endpoint_scope(EndpointScope::Yank)
+        .for_crate(&krate.name)
+        .check(req, conn)?;
+
+    state
+        .rate_limiter
+        .check_rate_limit(auth.user_id(), LimitedAction::YankUnyank, conn)?;
```

Comment:
If I understand correctly, the current implementation will not trigger a rate limit with invalid data. However, I believe we should check for a rate limit first, as data validation can (and should?) also be performed on the frontend before making a request. This can also prevent potential malicious requests.

Reply:
Regarding the version part, I think both options are acceptable, as an invalid version string will result in a 404 page. 
However, for invalid payload, as I previously mentioned, I recommend consuming the rate limit token.

Reply:
I just noticed that this function is used in two places, I'm wondering if it would be better to have something similar to the following:

```rust
enum YankAction {
    UpdateWithReason {
        yanked: Option<bool>,
        yank_message: Option<String>,
    },
    Toggle {
        yanked: bool,
    },
}

impl YankAction {
    pub fn from_update(data: &VersionUpdate) -> Self {
        ...
    }

    fn validate_with_version(&self, version: &Version) -> AppResult<()> {
        ...
    }

    fn apply(
        &self,
        state: &AppState,
        req: &Parts,
        conn: &mut impl Conn,
        version: &mut Version,
        krate: &Crate,
    ) -> AppResult<()> {
      ...
    }
}
```

I'm not sure if this would overcomplicate things, but IMHO, with this approach, we could distinguish the operations more easily, making it easier to follow and debug. Additionally, it would provide the potential opportunity to optimize each case individually while still sharing most of the functionality.

(Note: this is just a brief example. The naming, methods, and their arguments can be further modified as needed.)

Reply:
> However, for invalid payload, as I previously mentioned, I recommend consuming the rate limit token.

From our current API design, it seems we prefer do it before the auth check.

https://github.com/rust-lang/crates.io/blob/5f8fbb9c920a7fc346c7acf556110963b251efd8/src/controllers/krate/publish.rs#L62

Reply:
> I'm not sure if this would overcomplicate things, but IMHO, with this approach, we could distinguish the operations more easily, making it easier to follow and debug.

I just simplified the API design to make all updates more obvious. Directly pass the 'yanked' and 'yank_message' to the function, instead of implicitly depending on the updated behavior of the 'version' parameter. What do you think?

Reply:
> I believe we should check for a rate limit first, as data validation can (and should?) also be performed on the frontend before making a request. This can also prevent potential malicious requests.

I'm not sure whether I agree or not. It sounds reasonable at first, but our current rate limiting implementation performs SQL queries, and parsing a JSON payload should be quite a bit quicker than performing an SQL query. if we performed the rate limiting first then the API could be flooded with invalid data which would still take away a lot of SQL connections from the pool.

in any case this PR just continues with the same order that the other API endpoints are currently using so I think it's okay for now. we can probably (continue to) discuss and potentially address this in a dedicated issue/PR/Zulip thread.